Spigot-Server-Patches/0267-PlayerAdvancementGrantEvent.patch

```diff
@@ -0,0 +1,28 @@
+From c635468c1084f7ebd24ce0a50394cf6a9a009e81 Mon Sep 17 00:00:00 2001
+From: BillyGalbreath <Blake.Galbreath@GMail.com>
+Date: Fri, 19 Jan 2018 08:15:29 -0600
+Subject: [PATCH] PlayerAdvancementGrantEvent
+
+
+diff --git a/src/main/java/net/minecraft/server/AdvancementDataPlayer.java b/src/main/java/net/minecraft/server/AdvancementDataPlayer.java
+index 6896b709..be2b8848 100644
+--- a/src/main/java/net/minecraft/server/AdvancementDataPlayer.java
++++ b/src/main/java/net/minecraft/server/AdvancementDataPlayer.java
+@@ -198,6 +198,14 @@ public class AdvancementDataPlayer {
+         if (advancementprogress.a(s)) {
+             this.d(advancement);
+             this.i.add(advancement);
++            // Paper start
++            if (!new com.destroystokyo.paper.event.player.PlayerAdvancementGrantEvent(this.player.getBukkitEntity(), advancement.bukkit, s).callEvent()) {
```

Comment:
Looking at the code, this looks like it should be done 2 lines up and skip calling this.d/this.i.add if cancelled.

A lot of stuff happens in this.d, and it doesnt make sense to do it then revert it for this case.

So unless you can explain why that won't work, please move it up then looks good to me.

Reply:
I agree, which is why my own server added a Pre event even farther up than that. https://gyazo.com/94248e98ceab2a265ccc6b51ba05a8f8

I got the logic from a Forge PR https://github.com/MinecraftForge/MinecraftForge/pull/4257

I dont actually use this version, in my fork I put a Pre event much farther up https://gyazo.com/94248e98ceab2a265ccc6b51ba05a8f8 I left it here in the PR because of what I saw in the Forge PR. I figured it unregistered the world listeners and then re-added them if cancelled for a reason.

Give me a bit and I'll make the requested change ^_^

Reply:
Done.